finality-aleph/src/session_map.rs

```diff
@@ -0,0 +1,465 @@
+{
+    pub fn new(client: Arc<C>) -> Self {
+        Self {
+            client,
+            _phantom: PhantomData,
+        }
+    }
+}
+
+impl<C, B, BE> FinalityNotificator<FinalityNotification<B>, NumberFor<B>>
+    for FinalityNotificatorImpl<C, B, BE>
+where
+    C: ClientForAleph<B, BE> + Send + Sync + 'static,
+    C::Api: aleph_primitives::AlephSessionApi<B>,
+    B: Block,
+    BE: Backend<B> + 'static,
+{
+    fn notification_stream(&mut self) -> TracingUnboundedReceiver<FinalityNotification<B>> {
+        self.client.finality_notification_stream()
+    }
+
+    fn last_finalized(&self) -> NumberFor<B> {
+        self.client.info().finalized_number
+    }
+}
+
+#[derive(Clone)]
+/// Wrapper around Mapping from sessionId to Vec of AuthorityIds allowing mutation
+/// and hiding locking details
+struct SharedSessionMap(Arc<Mutex<SessionMap>>);
+
+#[derive(Clone)]
+/// Wrapper around Mapping from sessionId to Vec of AuthorityIds allowing only reads
+pub struct ReadOnlySessionMap {
+    inner: Arc<Mutex<SessionMap>>,
+}
+
+impl SharedSessionMap {
+    fn new() -> Self {
+        Self(Arc::new(Mutex::new(HashMap::new())))
+    }
+
+    async fn update(
+        &mut self,
+        id: SessionId,
+        authorities: Vec<AuthorityId>,
+    ) -> Option<Vec<AuthorityId>> {
+        self.0.lock().await.insert(id, authorities)
+    }
+
+    async fn prune_below(&mut self, id: SessionId) {
+        self.0.lock().await.retain(|&s, _| s >= id);
+    }
+
+    fn read_only(&self) -> ReadOnlySessionMap {
+        ReadOnlySessionMap {
+            inner: self.0.clone(),
+        }
+    }
+}
+
+impl ReadOnlySessionMap {
+    pub async fn get(&self, id: SessionId) -> Option<Vec<AuthorityId>> {
+        self.inner.lock().await.get(&id).cloned()
+    }
+}
+
+fn get_authorities_for_session<AP, B>(
+    authority_provider: &AP,
+    session_id: SessionId,
+    first_block: NumberFor<B>,
+) -> Vec<AuthorityId>
+where
+    B: Block,
+    AP: AuthorityProvider<NumberFor<B>>,
+{
+    if session_id == SessionId(0) {
+        authority_provider
+            .authorities(<NumberFor<B>>::saturated_from(0u32))
+            .expect("Authorities for the session 0 must be available from the beginning")
+    } else {
+        authority_provider.next_authorities(first_block).unwrap_or_else(||
+            panic!("Authorities for next session {:?} must be available at first block #{:?} of current session", session_id.0, first_block)
+        )
+    }
+}
+
+/// Returns None if the num is not a first block of some session otherwise returns Some(id) where
+/// id is id of a session that block starts.
+fn is_first_block<B: Block>(num: NumberFor<B>, period: SessionPeriod) -> Option<SessionId> {
+    let session = session_id_from_block_num::<B>(num, period);
+
+    if first_block_of_session::<B>(session, period) == num {
+        Some(session)
+    } else {
+        None
+    }
+}
+
+/// Struct responsible for updating session map
+pub struct SessionMapUpdater<AP, FN, B>
```

Comment:
I was thinking if logic in this struct can be a part of `SessionMap`? This is because normally we encapsulate a class behavior in the class itself, not in a separate class. I understand we have a distinction for `SharedSessionMap` and `ReadOnlySessionMap`, hence the need for another class. I'm just thinking of justification for having three separate structs. Alternative would be to 
* rename `SessionMapUpdater` to `SessionMap`, and incorporate current `SharedSessionMap` logic into it,
* do not distinguish on `SharedSessionMap` and `ReadOnlySessionMap`, have only one type `SessionMap` and have all the mutating/non-mutating logic encapsulated in renamed `SessionMap` struct

Reply:
Now with the additional subscriptions to insertions to the map, I think it makes sense to have such a struct (so the updater does not need to deal with it and only with the 'updating' logic) Wdyt?

Reply:
Yup, looks good, thanks.